lib/IDE/CodeCompletion.cpp

```diff
@@ -5318,6 +5318,13 @@ void CodeCompletionCallbacksImpl::doneParsing() {
     if (isDynamicLookup(*ExprType))
       Lookup.setIsDynamicLookup();
 
+    CodeCompletionResultBuilder Builder(CompletionContext.getResultSink(),
+                                        CodeCompletionResult::ResultKind::Keyword,
+                                        SemanticContextKind::CurrentNominal, {});
+    Builder.setKeywordKind(CodeCompletionKeywordKind::kw_self);
+    Builder.addTextChunk("self");
+    Builder.addTypeAnnotation(ExprType->getString());
```

Comment:
This code path is also used for completing  after`ModuleName.`, where `self` is not correct.  If you move this into `getValueExprCompletions` after the code that fixes `ExprType`, you can check whether `ExprType` is a module (see `tryModuleCompletions` for the check).  There should also be a test for this.

Thanks for working on this!

Reply:
Thanks for the hint. By the way, Is is ok to show `self` completions after `self` or instances? 

I'm asking because although this isn't wrong in terms of the language, it is either senseless in practice `(self.self...)` or something a user would very rarely do for some mysterious reason.

Also, the two `CompletionKind` that currently should show `self` and `Type` are `TypeIdentifierWithDot` and `TypeIdentifierWithoutDot`. I was a bit confused when I first discovered this. What cases do they cover?

Reply:
> I'm asking because although this isn't wrong in terms of the language, it is either senseless in practice (self.self...) or something a user would very rarely do for some mysterious reason.

Showing it is fine.

> Also, the two CompletionKind that currently should show self and Type are TypeIdentifierWithDot and TypeIdentifierWithoutDot. I was a bit confused when I first discovered this. What cases do they cover?

This is when we know syntactically that the LHS is a type name.  These are not the only cases you would want to handle because they won't include typenames in expressions.
```
let a: Foo<TypeIdentifierWithoutDot >
let b: Foo.<TypeIdentifierWithDot >
_ = Foo.<Neither of the above>
```

Reply:
I see. I noticed we don't have a `Type` completion for appropriate dot expressions as well.

> If you move this into getValueExprCompletions after the code that fixes ExprType, you can check whether ExprType is a module (see tryModuleCompletions for the check). There should also be a test for this.

I'm not experienced enough to understand what precisely happens in the "This is horrible" section, but it looks like it waits for generic types, so I suppose it doesn't affect modules whatsoever. That said and taking into account that `getValueExprCompletions` is invoked from several other paths apart from "DotExpr", wouldn't it be better to keep everything where it is? Otherwise I would have to check if we're dealing with `dotExpr` and write specific logic in a more general function that shouldn't care about that:

https://github.com/apple/swift/blob/19d8dd43fb29228e424550e14ba39a12fc6283d2/lib/IDE/CodeCompletion.cpp#L3255

Reply:
Fair enough, it doesn't actually matter if you do the check before or after that "horrible" section.  But you don't just want to handle DotExpr; for example, a postfix expression without a dot:
```
Foo <HERE>
```

Should still show `self`, but it should also add the leading ".".  This can be done by calling `addLeadingDot` which will add the dot only if it's needed.

Reply:
Handled those as well. I've tidied up the code but still haven't moved anything because otherwise I ended up touching paren expressions. I can change that, but there would have to be a way to check for `completionKind` in `getValueExprCompletions`.